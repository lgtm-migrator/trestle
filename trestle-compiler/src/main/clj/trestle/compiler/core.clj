(ns trestle.compiler.core
  (:require [clojure.spec.alpha :as s]
            [cheshire.core :as c]))

(s/def
  ::version
  (s/and pos-int? #(>= 1 %)))

(s/def ::dataset string?)
(s/def ::name string?)
(s/def ::type string?)
(s/def ::spatial boolean?)
(s/def ::identifier boolean?)
(s/def ::temporal string?)

; This isn't working, I want to enforce that we can only have a single qualifier on each fact
; Not sure how to do that
; I'd also like to ensure we have a maximum of 1 ID or spatial value
(s/def ::fact-qualifiers (s/or :spatial ::spatial :identifier ::identifier :temporal ::temporal))

;(s/def ::fact-qualifiers (s/or
;                           (s/keys :req-un [:spatial ::spatial])
;                           (s/keys :req-un [:identifier ::identifier])
;                           (s/keys :req-un [:temporal ::temporal])))

(s/def ::fact-params (s/keys
                       :req-un [::name ::type]
                       :opt-un [::fact-qualifiers]))

(s/def ::fact (s/merge ::fact-qualifiers ::fact-params))

(s/def ::facts (s/coll-of ::fact))

(s/def :dataset/spec
  (s/keys :req-un [::version ::dataset ::facts]))

(defn parse-json [file]
  (c/parse-stream (clojure.java.io/reader file) true))

(s/fdef parse-json
        :args (s/cat :file string?)
        :ret ::dataset)