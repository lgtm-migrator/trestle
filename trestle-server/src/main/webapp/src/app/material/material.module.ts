import {NgModule} from '@angular/core';
import {MatAutocompleteModule} from '@angular/material/autocomplete';
import {MatButtonToggleModule} from '@angular/material/button-toggle';
import {MatExpansionModule} from '@angular/material/expansion';
import {CdkTableModule} from '@angular/cdk/table';
import {MatButtonModule} from '@angular/material/button';
import {MatCardModule} from '@angular/material/card';
import {MatCheckboxModule} from '@angular/material/checkbox';
import {MatChipsModule} from '@angular/material/chips';
import {MatDialogModule} from '@angular/material/dialog';
import {MatFormFieldModule} from '@angular/material/form-field';
import {MatGridListModule} from '@angular/material/grid-list';
import {MatIconModule} from '@angular/material/icon';
import {MatInputModule} from '@angular/material/input';
import {MatListModule} from '@angular/material/list';
import {MatMenuModule} from '@angular/material/menu';
import {MatOptionModule} from '@angular/material/core';
import {MatProgressBarModule} from '@angular/material/progress-bar';
import {MatProgressSpinnerModule} from '@angular/material/progress-spinner';
import {MatRadioModule} from '@angular/material/radio';
import {MatSelectModule} from '@angular/material/select';
import {MatSidenavModule} from '@angular/material/sidenav';
import {MatSliderModule} from '@angular/material/slider';
import {MatSlideToggleModule} from '@angular/material/slide-toggle';
import {MatTableModule} from '@angular/material/table';
import {MatTabsModule} from '@angular/material/tabs';
import {MatToolbarModule} from '@angular/material/toolbar';
import {MatTooltipModule} from '@angular/material/tooltip';
import {MatPaginatorModule} from '@angular/material/paginator';

@NgModule({
  imports: [
    CdkTableModule,
    MatAutocompleteModule,
    MatButtonModule,
    MatButtonToggleModule,
    MatCardModule,
    MatCheckboxModule,
    MatChipsModule,
    MatDialogModule,
    MatExpansionModule,
    MatFormFieldModule,
    MatGridListModule,
    MatIconModule,
    MatInputModule,
    MatListModule,
    MatMenuModule,
    MatOptionModule,
    MatPaginatorModule,
    MatProgressBarModule,
    MatProgressSpinnerModule,
    MatRadioModule,
    MatSelectModule,
    MatSidenavModule,
    MatSliderModule,
    MatSlideToggleModule,
    MatTableModule,
    MatTabsModule,
    MatToolbarModule,
    MatTooltipModule],
  exports: [
    CdkTableModule,
    MatAutocompleteModule,
    MatButtonModule,
    MatButtonToggleModule,
    MatCardModule,
    MatCheckboxModule,
    MatChipsModule,
    MatDialogModule,
    MatExpansionModule,
    MatFormFieldModule,
    MatGridListModule,
    MatIconModule,
    MatInputModule,
    MatListModule,
    MatMenuModule,
    MatOptionModule,
    MatPaginatorModule,
    MatProgressBarModule,
    MatProgressSpinnerModule,
    MatRadioModule,
    MatSelectModule,
    MatSidenavModule,
    MatSliderModule,
    MatSlideToggleModule,
    MatTableModule,
    MatTabsModule,
    MatToolbarModule,
    MatTooltipModule]
})
export class MaterialModule {
}
