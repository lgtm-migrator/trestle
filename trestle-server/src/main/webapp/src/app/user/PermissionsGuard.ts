/**
 * Created by nrobison on 1/20/17.
 */
import {ActivatedRouteSnapshot, CanActivate, RouterStateSnapshot} from '@angular/router';
import {Observable} from 'rxjs';
import {Injectable} from '@angular/core';
import {State} from '../reducers';
import {select, Store} from '@ngrx/store';
import {Privileges, TrestleUser} from './trestle-user';

@Injectable()
export class PermissionsGuard implements CanActivate {

  private user: TrestleUser;
  private subscription;

  public constructor(private store: Store<State>) {
    this.subscription = this.store.pipe(select('user'), select('user')).subscribe(user => this.user = user);
  }

  public canActivate(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Observable<boolean> | Promise<boolean> | boolean {
    const roles = route.data['roles'] as Privileges[];
    return this.user.hasRequiredPrivileges(roles);
  }
}
