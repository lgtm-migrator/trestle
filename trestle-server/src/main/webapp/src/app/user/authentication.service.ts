/**
 * Created by nrobison on 1/19/17.
 */
import {Injectable} from '@angular/core';
import {Router} from '@angular/router';
import {Privileges, TrestleUser} from './trestle-user';
import {HttpClient} from '@angular/common/http';
import {JwtHelperService} from '@auth0/angular-jwt';
import {environment} from '../../environments/environment';
import {Observable} from 'rxjs';
import {map} from 'rxjs/operators';

const _key: string = 'access_token';

export class TrestleToken {

  private readonly exp: number;
  private readonly iat: number;
  private readonly user: ITrestleUser;

  public constructor(token: any) {
    this.exp = token['exp'];
    this.iat = token['iat'];
    this.user = JSON.parse(token['data4j']);
  }

  /**
   * Get token expiration time (in ms from Unix epoch)
   * @returns {number}
   */
  public getExpiration() {
    return this.exp;
  }

  /**
   * Get issuance time (in ms from Unix epoch)
   * @returns {number}
   */
  public getIssue() {
    return this.iat;
  }

  /**
   * Get the user from the token
   * @returns {ITrestleUser}
   */
  public getUser() {
    return this.user;
  }
}

export interface ITrestleUser {
  id?: number;
  firstName: string;
  lastName: string;
  username: string;
  email: string;
  password: string;
  privileges: number;
}

@Injectable()
export class AuthService {

  private jwtHelper: JwtHelperService;
  private readonly baseUrl;

  constructor(private router: Router, private http: HttpClient) {
    this.jwtHelper = new JwtHelperService();
    this.baseUrl = environment.baseUrl;
  }

  /**
   * Attempt to login the given use
   * @param {string} username
   * @param {string} password
   */
  public login(username: string, password: string): Observable<{ user: TrestleUser, token: string }> {
    return this.http.post(this.baseUrl + '/auth/login', {username, password: password}, {
      responseType: 'text'
    })
      .pipe(map(token => {
        const t = this.getToken2(token);
        return {user: new TrestleUser(t.getUser()), token}
      }));
  }

  /**
   * Logout the user
   */
  public logout(): Observable<string> {
    console.debug("Logging out");
    return this.http.post(this.baseUrl + '/auth/logout', null, {
      responseType: 'text'
    });
  }

  /**
   * Is the user currently logged in?
   * @returns {boolean}
   */
  public loggedIn(): boolean {
    const token = localStorage.getItem(_key);
    if (token) {
      console.debug(
        this.jwtHelper.decodeToken(token),
        this.jwtHelper.getTokenExpirationDate(token),
        this.jwtHelper.isTokenExpired(token)
      );
    }
    return !this.jwtHelper.isTokenExpired(token);
  }

  /**
   * Does the user have AT LEAST Admin permissions?
   * @returns {boolean}
   */
  public isAdmin(): boolean {
    if (this.loggedIn()) {
      const token = this.getToken();
      // tslint:disable-next-line:no-bitwise
      return (token !== null) && ((token.getUser().privileges & Privileges.ADMIN) > 0);
    }
    return false;
  }

  /**
   * Get the user, if one exists
   * @returns {TrestleUser | null}
   */
  public getUser(): TrestleUser | null {
    if (this.loggedIn()) {
      const token = this.getToken();
      if (token) {
        return new TrestleUser(token.getUser());
      }
    }
    console.error('User is not logged in');
    return null;
  }

  /**
   * Determines if the logged-in user has the necessary roles to perform a certain function
   * @param roles - Array of required Privileges
   * @returns {boolean} - has all the required roles
   */
  public hasRequiredRoles(roles: Privileges[]): boolean {
    const user = this.getUser();
    if (user == null) {
      return false;
    }

    return user.hasRequiredPrivileges(roles);
  }

  public getEncodedToken(): string | null {
    return localStorage.getItem(_key);
  }

  private getToken(): TrestleToken | null {
    const jwtToken = this.getEncodedToken();
    if (jwtToken) {
      return new TrestleToken(this.jwtHelper.decodeToken(jwtToken));
    }
    return null;
  }

  private getToken2 = (jwtToken: string): TrestleToken => {
    return new TrestleToken(this.jwtHelper.decodeToken(jwtToken));
  };
}
