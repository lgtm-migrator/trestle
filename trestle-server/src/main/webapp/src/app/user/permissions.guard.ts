/**
 * Created by nrobison on 1/20/17.
 */
import {ActivatedRouteSnapshot, CanActivate, RouterStateSnapshot} from '@angular/router';
import {Observable} from 'rxjs';
import {Injectable} from '@angular/core';
import {selectUserFromUser, State} from '../reducers';
import {select, Store} from '@ngrx/store';
import {Privileges, TrestleUser} from './trestle-user';

@Injectable()
export class PermissionsGuard implements CanActivate {

  private user: TrestleUser;
  private subscription;

  public constructor(private store: Store<State>) {
    this.subscription = this.store.pipe(select(selectUserFromUser));
  }

  public canActivate(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Observable<boolean> | Promise<boolean> | boolean {
    const roles = route.data['roles'] as Privileges[];
    return this.user.hasRequiredPrivileges(roles);
  }
}
