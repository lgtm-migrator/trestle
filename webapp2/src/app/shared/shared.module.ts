import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import {IndividualService} from './individual/individual.service';
import {DATASET_CACHE, DatasetService} from './dataset/dataset.service';
import {CacheService} from './cache/cache.service';
import {ArraySortPipePipe} from './pipes/array-sort.pipe';
import {DATASET_CACHE_DI_CONFIG} from './shared.config';



@NgModule({
  declarations: [ArraySortPipePipe],
  providers: [
    IndividualService,
    DatasetService,
    CacheService,
    {
      provide: DATASET_CACHE, useFactory: () => (new CacheService<string, string[]>(DATASET_CACHE_DI_CONFIG))
    }],
  imports: [
    CommonModule
  ]
})
export class SharedModule { }
