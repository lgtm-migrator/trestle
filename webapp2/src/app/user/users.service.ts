/**
 * Created by nrobison on 1/19/17.
 */
import {Injectable} from "@angular/core";
import {Observable} from "rxjs";
import {ITrestleUser} from "./authentication.service";
import {HttpClient} from "@angular/common/http";
import {map} from "rxjs/operators";

@Injectable()
export class UserService {
  constructor(private trestleHttp: HttpClient) {
  }

  /**
   * Get a list of all registered users
   * Requires ADMIN permissions
   *
   * @returns {Observable<R>}
   */
  public getUsers(): Observable<ITrestleUser[]> {
    return this.trestleHttp.get<ITrestleUser[]>("/users");
  }

  /**
   * Sends an {@link ITrestleUser} object to the server
   * Either creates the new user or modifies an existing one
   *
   * @param user - User to create/modify
   * @returns {Observable<R>}
   */
  public modifyUser(user: ITrestleUser): Observable<any> {
    return this.trestleHttp.post("/users", user);
  }

  /**
   * Delete user from database
   * @param id - ID of user to delete
   * @returns {Observable<R>}
   */
  public deleteUser(id: number): Observable<any> {
    return this.trestleHttp.delete("/users/" + id);
  }

  /**
   * Determines whether or not a given username exists in the application
   *
   * @param {string} username
   * @returns {Observable<boolean>}
   */
  public userExists(username: string): Observable<boolean> {
    console.debug("Checking username:", username);
    return this.trestleHttp.get<string>("/users/exists/" + username)
      .pipe(map(res => res === "true"));
  }
}
