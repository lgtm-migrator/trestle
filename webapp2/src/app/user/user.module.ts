import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import {HttpClientModule} from "@angular/common/http";
import {JwtModule} from "@auth0/angular-jwt";
import {AuthService} from "./authentication.service";
import {LoggedInGuard} from "./LoggedInGuard";
import {PermissionsGuard} from "./PermissionsGuard";
import {DefaultRouteGuard} from "./DefaultRouteGuard";
import {UserService} from "./users.service";

const _key: string = "access_token";


@NgModule({
  declarations: [],
  imports: [
    HttpClientModule,
    CommonModule,
    JwtModule.forRoot({
      config: {
        tokenGetter: tokenGetter
      }
    })
  ],
  providers: [
    AuthService,
    LoggedInGuard,
    PermissionsGuard,
    DefaultRouteGuard,
    UserService
  ]
})
export class UserModule { }

export function tokenGetter() {
  return localStorage.getItem(_key);
}
