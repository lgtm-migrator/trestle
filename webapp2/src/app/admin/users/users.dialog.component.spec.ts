import {async, ComponentFixture, TestBed} from '@angular/core/testing';
import {UserDialogComponent} from './users.dialog.component';
import {MaterialModule} from '../../material/material.module';
import {UserModule} from '../../user/user.module';
import {ReactiveFormsModule} from '@angular/forms';
import {SharedModule} from '../../shared/shared.module';
import {NO_ERRORS_SCHEMA} from '@angular/core';
import {HttpClientTestingModule} from '@angular/common/http/testing';
import {MatDialog, MatDialogRef} from '@angular/material/dialog';
import {of} from 'rxjs';

class MatDialogMock {
  open() {
    return {
      afterClosed: () => of()
    };
  }
};

describe('UsersDialogComponent', () => {
  let component: UserDialogComponent;
  let fixture: ComponentFixture<UserDialogComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [MaterialModule, UserModule, ReactiveFormsModule, SharedModule, HttpClientTestingModule],
      providers: [{
        provide: MatDialogRef, useClass: MatDialogMock
      }],
      declarations: [UserDialogComponent],
      schemas: [NO_ERRORS_SCHEMA]
    })
      .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(UserDialogComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });
});
