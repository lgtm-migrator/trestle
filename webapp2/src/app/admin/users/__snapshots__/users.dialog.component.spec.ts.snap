// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersDialogComponent should create 1`] = `undefined`;
